src/weak_memory.rs

```diff
@@ -0,0 +1,271 @@
+                    return true;
+                }
+
+                // The current non-SC load can only read-from the latest SC store (or any stores later than that
+                // SC store)
+                if store_elem.timestamp <= clocks.write_seqcst[store_elem.store_index]
+                    && store_elem.is_seqcst
+                {
+                    log::info!("Stopped due to needing to load from the last SC store");
+                    keep_searching = false;
+                    return true;
+                }
+
+                // The current SC load can only read-from the last store sequenced-before
+                // the last SC fence (or any stores later than the SC fence)
+                if is_seqcst && store_elem.timestamp <= clocks.read_seqcst[store_elem.store_index] {
+                    log::info!("Stopped due to sc load needing to load from the last SC store before an SC fence");
+                    keep_searching = false;
+                    return true;
+                }
+
+                true
+            })
+            .filter(|&store_elem| {
+                if is_seqcst {
+                    // An SC load needs to ignore all but last store maked SC (stores not marked SC are not
+                    // affected)
+                    let include = !(store_elem.is_seqcst && found_sc);
+                    found_sc |= store_elem.is_seqcst;
+                    include
+                } else {
+                    true
+                }
+            });
+
+        candidates.choose(rng)
+    }
+
+    /// ATOMIC STORE IMPL in the paper
+    /// The paper also wants the variable's clock (AtomicMemoryCellClocks::sync_vector in our code)
+    /// to be in the store element, but it's not used anywhere so we don't actually need it
+    fn store_impl(
+        &mut self,
+        val: ScalarMaybeUninit<Tag>,
+        index: VectorIdx,
+        thread_clock: &VClock,
+        is_seqcst: bool,
+    ) {
+        let store_elem = StoreElement {
+            store_index: index,
+            timestamp: thread_clock[index],
+            // In the language provided in the paper, an atomic store takes the value from a
+            // non-atomic memory location.
+            // But we already have the immediate value here so we don't need to do the memory
+            // access
+            val,
+            is_seqcst,
+            loads: RefCell::new(FxHashMap::default()),
+        };
+        self.buffer.push_back(store_elem);
+        if self.buffer.len() > STORE_BUFFER_LIMIT {
+            self.buffer.pop_front();
+        }
+        if is_seqcst {
+            // Every store that happens before this needs to be marked as SC
+            // so that in fetch_store, only the last SC store (i.e. this one) or stores that
+            // aren't ordered by hb with the last SC is picked.
+            //
+            // This is described in the paper (§4.5) and implemented in tsan11
```

Comment:
Nothing in this file even says which paper you refer to; would be good to add a reference.